Adds gRPC Server-side TLS Support

danehans opened the pull request:
Adds gRPC Server-side TLS support to Fortio grpc client and server. Addresses `TODO: option to specify certs.` in `fgrpc/grpcrunner.go`.

Commit:
Adds gRPC mTLS Support

ldemailly commented:
I don't think we should put self signed certs in the release docker image

maybe it can be a volume/mount like the json data dir ?

for testing we can have indeed a script generating a sample

danehans commented:
Thank you for taking a look at the PR. 

> I don't think we should put self signed certs in the release docker image

Agreed, that is why I did not update release/Dockerfile.

> maybe it can be a volume/mount like the json data dir ?

I'm not familiar with how the json data dir is mounted. Do you have a pointer?

> for testing we can have indeed a script generating a sample

Agreed, I wanted to make sure we can perform e2e testing on the feature. This is why I updated the non-release Docker image. Maybe I don't understand how the non-release and release images work with one another.

ldemailly commented:
> Thank you for taking a look at the PR.

I haven't yet, really. will now

> I'm not familiar with how the json data dir is mounted. Do you have a pointer?

https://github.com/istio/fortio/blob/master/Dockerfile#L33

> Agreed, I wanted to make sure we can perform e2e testing on the feature. This is why I updated the non-release Docker image. Maybe I don't understand how the non-release and release images work with one another.

Dockerfile is the release for istio/fortio
there is additional code in release/ that extract a tgz out of it but it's secondary

ldemailly commented on Dockerfile:
```diff
@@ -20,10 +20,17 @@ RUN /usr/local/go/bin/go version
 RUN CGO_ENABLED=0 GOOS=darwin /usr/local/go/bin/go build -a -ldflags -s -o fortio.go18.mac istio.io/fortio
 # Just check it stays compiling on Windows (would need to set the rsrcDir too)
 RUN CGO_ENABLED=0 GOOS=windows go build -a -o fortio.exe istio.io/fortio
+# Generate TLS assets for secure grpc
```
undo all this

ldemailly commented on ftls/ftls.go:
```diff
@@ -0,0 +1,62 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ftls
```
I don't think we need a brand new top level package just for this

it can be a file inside fnet

ldemailly commented on ftls/ftls.go:
```diff
@@ -0,0 +1,62 @@
+
+// NewCertPool creates an x509 certPool with the provided CA files.
+func NewCertPool(CAFiles []string) (*x509.CertPool, error) {
+	certPool := x509.NewCertPool()
```
this seems overly complicated for a testing tool we only need
https://godoc.org/google.golang.org/grpc/credentials#NewClientTLSFromFile
no ?

(though maybe it is worth having this for http tls too)

danehans replied:
I thought this would be needed to setup one or more trusted CA for authorizing client/server connections.

ldemailly commented on fgrpc/grpcrunner.go:
```diff
@@ -41,11 +42,22 @@ const (
+		}
+		creds := credentials.NewTLS(tlsCfg)
+		opts = grpc.WithTransportCredentials(creds)
+	case strings.HasPrefix(serverAddr, "https://"):
```
I don't think the switch between standard CA vs custom CA should be based on https:// prefix
it's a config option, to be added to grpcoptions

ldemailly commented on fgrpc/pingsrv.go:
```diff
@@ -51,9 +53,19 @@ func (s *pingSrv) Ping(c context.Context, in *PingMessage) (*PingMessage, error)
-	grpcServer := grpc.NewServer()
+	var opts []grpc.ServerOption
+	if secure {
+		tlsCfg, err := ftls.NewCredentials(false, ftls.DefaultServerCert, ftls.DefaultServerKey,
```
if they are hard coded what is the point of passing constants.
we need some grpcoptions struct to control tls instead of just 1 boolean

ldemailly commented on ftls/info.go:
```diff
@@ -0,0 +1,98 @@
+// ClientConfig returns a tls.Config for client use.
+func (info *TLSInfo) ClientConfig() (*tls.Config, error) {
+	// CA for verifying the server
+	pool, err := NewCertPool([]string{info.CAFile})
```
how would you put both the standard CA and additional one in 1 config ?

danehans replied:
Change `CAFile` to be of type `[]string` that represent one or more paths to trusted CA certs. Provide the slice to `NewCertPool`. Create a cmdline flag that allows a user to specify multiple comma-separated cert paths. Is that acceptable?

ldemailly replied:
i'd rather have just the NewClientTLSFromFile for now I think; and less code

ldemailly reviewed (commented):
tbh in addition to all the technical issues/details below,

I'm really not a fan of adding all this code - it might grow on me though once simplified and working but... not sure.... I'll give it some more thoughts

Commit:
Addresses @ldemailly 3/15/18 feedback

Commit:
Move cert log msgs and fix typo

danehans commented:
@ldemailly Thanks for your review. I addressed your comments and have updated the PR. The purpose of this PR is to address `TODO: option to specify certs`. I should have created an issue to discuss my intentions before getting this far. If my approach is unwanted, let me know how you would like to address the TODO.

danehans commented:
OK. I'll strip most of this stuff out and update the PR.

Commit:
Addresses round 2 of @ldemailly comments

ldemailly commented:
Sorry been crazy busy... will look more soon

Commit:
Added docs and minor fixups

Commit:
Updated Dial return error

Commit:
Updates 2nd ping server fail test with new params

Commit:
Removes unneeded var declaration in Dial func

ldemailly commented on README.md:
```diff
@@ -77,8 +77,12 @@ target is a url (http load tests) or host:port (grpc health test).  flags are:
   -grpc-port string
 	grpc server port. Can be in the form of host:port, ip:port or port.
 	(default "8079")
-  -grpc-secure
```
we still need a way to standard TLS no?

danehans replied:
This PR provides backwards compatibility. Instead of passing the  `-grpc-secure` bool,  `Dial` will use the one of the following:

A. The cert from `-server-cert` if it's provided. I can easily add the flag back in, but IMO it no longer adds value.
B. Us `nil` creds if dest is an https prefix (this provides backwards compatibility)
C. Insecure Dial if A or C do not apply.

I thought this is a cleaner approach than setting the bool and cert/key flags. IMO having both sets of flags causes user confusion. Using the legacy bool flag made sense, a user flips the secure/insecure switch.

ldemailly replied:
I'm not following - how do I get the previous behavior exactly ?

danehans replied:
The previous behavior is accomplished by using an https prefix with the grpc destination.
For example:

__grpc client__
```
fortio grpcping https://fortio.istio.io
```
__or any client that uses `GRPCRunnerOptions`__
```
fortio curl -loglevel debug http://localhost:8080/fortio/fetch/localhost:8080/fortio/?url=https://fortio.istio.io:443&load=Start&qps=-1&json=on&n=100&runner=grpc
```

ldemailly replied:
ic I missed that - does it work with fortio load -grpc https://... too ? (it should I guess as destination is passed down unchanged if I recall)

we should put tls/no tls in the (json) result object

we should probably document that behavior somewhere (that grpc   host:port is insecure unless a cert is given and https://host:port is secure)

danehans replied:
> does it work with fortio load -grpc https://... too ? 

It does. I extensively tested this patch today. I documented the details [here](https://gist.github.com/danehans/694cb0ce68017a6e44aea85f7b47d2f7#file-fortio_mtls_test_cases-md).

> we should probably document that behavior somewhere (that grpc host:port is insecure unless a cert is given and https://host:port is secure)

I updated the Readme.

ldemailly commented on fortio_main.go:
```diff
@@ -70,15 +70,18 @@ func usage(msgs ...interface{}) {
+	goMaxProcsFlag  = flag.Int("gomaxprocs", 0, "Setting for runtime.GOMAXPROCS, <1 doesn't change the default")
+	profileFlag     = flag.String("profile", "", "write .cpu and .mem profiles to file")
+	grpcFlag        = flag.Bool("grpc", false, "Use GRPC (health check) for load testing")
+	serverCertFlag  = flag.String("server-cert", "",
```
add a new line to minimize the diff
(go can be annoying with reformatting to align - also as mentioned I think the grpc-secure flag is still needed, the cert,ca,key are optional)

danehans replied:
Will do. For clarity, the ca flag was dropped since refactoring this PR based on your initial feedback.

ldemailly commented on fortio_main.go:
```diff
@@ -162,7 +165,7 @@ func main() {
 	case "server":
 		isServer = true
-		fgrpc.PingServer(*grpcPortFlag, fgrpc.DefaultHealthServiceName)
+		fgrpc.PingServer(*grpcPortFlag, *serverCertFlag, *serverKeyFlag, fgrpc.DefaultHealthServiceName)
```
I'm thinking we should add some GRPC(Server)Options instead of 5 string params
(shared between the client and server maybe)
but maybe that can be a followup

danehans replied:
I would like to address that in a follow-on PR. This PR has been around for a while. I opened issue
https://github.com/istio/fortio/issues/220 and assigned myself.

ldemailly replied:
k, ty

ldemailly commented on fortio_main.go:
```diff
@@ -249,7 +252,7 @@ func fortioLoad(justCurl bool, percList []float64) {
 			RunnerOptions:      ro,
 			Destination:        url,
-			Secure:             *grpcSecureFlag,
+			Cert:               *serverCertFlag,
```
the 2 are different

danehans replied:
Understood. See my response above and let me know your thoughts. I want to make sure we're on the same page before I rebase this PR (it was a PITA the last time I rebased).

ldemailly replied:
k

ldemailly commented on ui/uihandler.go:
```diff
@@ -153,6 +153,7 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 	qps, _ := strconv.ParseFloat(r.FormValue("qps"), 64)      // nolint: gas
 	durStr := r.FormValue("t")
 	grpcSecure := (r.FormValue("grpc-secure") == "on")
+	cert := r.FormValue("cert")
```
for this I think we should just use the flags (if you want a fortio UI using different cert, start a new one with different cert path)

danehans replied:
From what I see, the challenge with that approach is:

1. `uihandler.go` creates instance `o` of the `fgrpc.GRPCRunnerOptions` object .
2. The flags created in `fortio_main.go` are not available to `uihandler.go` or any other pkg for that matter. Therefore, the `Cert` field in `o` can not be populated based on any flags.

I believe what you are asking for can be accomplished by putting Fortio flags into a pkg (i.e. cli ) and having `fortio_main.go` and any other dependent pkg's import `cli`. If this has your interest, please create an Issue, assign me and I'll work on refactoring the Fortio cli into a separate pkg. However, I don't think that work should block this PR.

ldemailly replied:
we could also pass those global options to ui.Serve()

danehans replied:
Yes, assuming the cli options are in a separate pkg. As I mentioned, I am happy to take on this Fortio cli refactoring. Just create an Issue and assign me. Until then, I think providing a field to enter the cert path is a reasonable option.

ldemailly reviewed (commented):
let's use --key  --cert  --cacert 
like curl
and
https://istio.io/docs/tasks/security/mutual-tls.html#testing-the-authentication-setup

danehans reviewed (commented):
Thanks for the follow-up review!

Commit:
Realigns flags in fortio_main

ldemailly commented on fgrpc/grpcrunner.go:
```diff
@@ -41,16 +41,28 @@ const (
+		log.Infof("Using server certificate %v to construct TLS credentials", cert)
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	case strings.HasPrefix(serverAddr, "https://"):
+		creds := credentials.NewTLS(nil)
```
ic this is where the https/tls behaviour is

danehans replied:
Yeah, all the intelligence is in `Dial`.

ldemailly commented on fortio_main.go:
```diff
@@ -312,20 +315,19 @@ func fortioLoad(justCurl bool, percList []float64) {
 
 func grpcClient() {
-	if len(flag.Args()) != 1 {
+	if len(flag.Args()) < 1 {
```
which case is that change for ?

ldemailly commented on fgrpc/pingsrv_test.go:
```diff
@@ -29,28 +30,53 @@ func init() {
+	iPort := PingServer("0", "", "", "foo")
+	iAddr := fmt.Sprintf("localhost:%d", iPort)
+	t.Logf("test insecure grpc ping server running, will connect to %s", iAddr)
+	sPort := PingServer("0", "../testdata/server.crt", "../testdata/server.key",
```
isn't this relative path making success dependent on where it is ran from ? (somehow doubt it works from both parent or different directories?)

not sure we need a new directory - tbd if go testing has provision for files relative to test files?

danehans replied:
I tested by copying the cert/key from the testdata directory to the project's root directory and updating the following in `grpcrunner_test.go`:
```
	svrCrt = "../server.crt"
	svrKey = "../server.key"
```

The tests passed.

> not sure we need a new directory - tbd if go testing has provision for files relative to test files?

I like the idea of storing all test data for a project in a specific directory. I think it's clean and makes it very apparent to devs. Let me know where I should move the test cert/key if you want me to remove the directory.

Commit:
Updates cert/key flags and updates readme

ldemailly commented:
this is good, do you have time to raise the coverage ? (for instance pass invalid credentials, connect to the wrong port/settings,...)

ldemailly commented:
where is -cacert ? I see in the example you pass a cert to the client, I assume that implies it expects that cert, but it should be a CA instead ?

ldemailly commented on README.md:
```diff
@@ -80,8 +80,12 @@ target is a url (http load tests) or host:port (grpc health test).  flags are:
-	Use secure transport (tls) for GRPC
+  -cert
+	Full path to the server certificate required for secure grpc.
+	Used by grpcping and server. (default "") means secure grpc is disabled.
```
add:

"unless https:// destination is used, in which case standard TLS is expected (with valid server certificates signed by the internet standard CAs)" 

or some such

ldemailly commented on fgrpc/grpcrunner.go:
```diff
@@ -41,16 +41,28 @@ const (
+	//var creds credentials.TransportCredentials
+	switch {
+	case cert != "":
+		creds, err := credentials.NewClientTLSFromFile(cert, "")
```
I wonder here if we're confusing client and server cert

we should be able to do mTLS by providing an optional client cert; and expect a server cert signed by a custom CA

danehans replied:
I updated the name of this PR to properly reflect the implemented functionality. This PR originally provided mTLS, but I simplified things based on your initial feedback. This PR will allow clients (i.e. grpcping) to authenticate the Fortio gRPC server and encrypt the data exchanged. In my original PR, the server cert was fetched from the CA (Fortio server). Since refactoring, the server cert must be shipped with the client-side application. [This](https://bbengfort.github.io/programmer/2017/03/03/secure-grpc.html) blog provides details of the original and current approach for this PR. Let me know how you would like to proceed.

danehans replied:
[This section](https://en.wikipedia.org/wiki/Transport_Layer_Security#Basic_TLS_handshake) provides a detailed explanation of the basic TLS handshake implemented by this PR.

Commit:
Adds more details to docs

danehans commented:
@ldemailly I want to make sure we're on the same page for next steps with this PR. Is test coverage the last outstanding item or is anything else missing?

ldemailly commented:
> I want to make sure we're on the same page for next steps with this PR. Is test coverage the last outstanding item or is anything else missing?

Can you add the CA support without a lot of changes ?
Bundling the server's cert with the client is strange/uncommon. Also did you check what happens if you for instance pass your test cert and hit a real cert like https://fortio.istio.io/ ?

danehans commented:
> Can you add the CA support without a lot of changes ?

The PR will require a fair bit of refactoring. At this point, I would prefer closing this out. If mTLS is a requirement, can you create a new issue describing the required functionality and assign it to me?

> Bundling the server's cert with the client is strange/uncommon.

Server-side TLS is less secure but is easier to implement. Options include A) Obtain a real cert for Fortio server so the cert does not need to get bundled with the client or B) Setting up a CA so the client can validate the cert provided by the server during the handshake. If we go with option B, we might as well implement mTLS.

> Also did you check what happens if you for instance pass your test cert and hit a real cert like https://fortio.istio.io/ ?

The test case you describe will fail. The client will use an incorrect public key from the test cert to encrypt messages to https://fortio.istio.io/ and the TLS handshake will fail. I could have `Dial` return nil, err if `cert` and a https prefix in `serverAddr` is provided:
```
$ grpcping -cert /etc/ssl/certs/server.crt https://fortio.istio.io:443
19:29:30 I grpcrunner.go:56> Using server certificate /etc/ssl/certs/server.crt to construct TLS credentials
19:29:30 I grpcrunner.go:207> stripping https scheme. grpc destination: fortio.istio.io:443. grpc port: 443
19:29:30 E pingsrv.go:97> grpc error from Ping0 rpc error: code = Unavailable desc = all SubConns are in TransientFailure
```
However, using `grpcping` with a cert provided by a valid CA and/or prefix `https://` will succeed:
```
# Using -cert flag and https prefix
$ grpcping -cert /etc/ssl/certs/ca-certificates.crt  https://fortio.istio.io:443
19:49:50 I grpcrunner.go:56> Using server certificate /etc/ssl/certs/ca-certificates.crt to construct TLS credentials
19:49:50 I grpcrunner.go:207> stripping https scheme. grpc destination: fortio.istio.io:443. grpc port: 443
19:49:51 I pingsrv.go:129> Ping RTT 40955113 (avg of 39416088, 42003584, 41445667 ns) clock skew 14440715
Clock skew histogram usec : count 1 avg 14440.715 +/- 0 min 14440.715 max 14440.715 sum 14440.715
# range, mid point, percentile, count
>= 14440.7 <= 14440.7 , 14440.7 , 100.00, 1
# target 50% 14440.7
RTT histogram usec : count 3 avg 40955.113 +/- 1112 min 39416.088 max 42003.584 sum 122865.339
# range, mid point, percentile, count
>= 39416.1 <= 40000 , 39708 , 33.33, 1
> 40000 <= 42003.6 , 41001.8 , 100.00, 2
# target 50% 40500.9

# Using -cert flag only
$ grpcping -cert /etc/ssl/certs/ca-certificates.crt fortio.istio.io:443
19:50:02 I grpcrunner.go:56> Using server certificate /etc/ssl/certs/ca-certificates.crt to construct TLS credentials
19:50:02 I pingsrv.go:129> Ping RTT 41519156 (avg of 38733801, 42954100, 42869569 ns) clock skew 15710098
Clock skew histogram usec : count 1 avg 15710.098 +/- 0 min 15710.098 max 15710.098 sum 15710.098
# range, mid point, percentile, count
>= 15710.1 <= 15710.1 , 15710.1 , 100.00, 1
# target 50% 15710.1
RTT histogram usec : count 3 avg 41519.157 +/- 1970 min 38733.801 max 42954.1 sum 124557.47
# range, mid point, percentile, count
>= 38733.8 <= 40000 , 39366.9 , 33.33, 1
> 40000 <= 42954.1 , 41477.1 , 100.00, 2
# target 50% 40738.5
```

Using https prefix only:
```
$ grpcping https://fortio.istio.io:443
19:51:56 I grpcrunner.go:207> stripping https scheme. grpc destination: fortio.istio.io:443. grpc port: 443
Clock skew histogram usec : count 1 avg 13694.372 +/- 0 min 13694.372 max 13694.372 sum 13694.372
# range, mid point, percentile, count
>= 13694.4 <= 13694.4 , 13694.4 , 100.00, 1
# target 50% 13694.4
RTT histogram usec : count 3 avg 41078.457 +/- 1155 min 39445.253 max 41908.073 sum 123235.372
# range, mid point, percentile, count
>= 39445.3 <= 40000 , 39722.6 , 33.33, 1
> 40000 <= 41908.1 , 40954 , 100.00, 2
19:51:56 I pingsrv.go:129> Ping RTT 41078457 (avg of 39445253, 41908073, 41882046 ns) clock skew 13694372
# target 50% 40477
```

Commit:
Updates e2e/ut, docs and Dial func to support vhost override

ldemailly commented:
> $ grpcping -cert /etc/ssl/certs/ca-certificates.crt fortio.istio.io:443

oh strange so that 'cert' argument is also cacert !

ldemailly commented:
can we split the 2 then, call cacert the one for client and cert for the server

Commit:
Implements new -ca-cert flag

danehans commented:
> can we split the 2 then, call cacert the one for client and cert for the server

Done in latest commit.

ldemailly commented on README.md:
```diff
@@ -81,14 +81,17 @@ target is a url (http load tests) or host:port (grpc health test).  flags are:
-	Full path to the key required for grpc TLS. Applicable to Fortio server.
+	Full path to the key required for grpc TLS. Applicable to the server command.
 	(default "") means no user-provided key.
+  -ca-cert
```
sorry to be a PITA but please use  cacert not ca-cert, trying to be command line compatible with curl's:
```
    --cacert FILE   CA certificate to verify peer against (SSL)
```

ldemailly replied:
( mentioned a few times, starting on https://github.com/istio/fortio/pull/204#pullrequestreview-107094841 )

ldemailly commented on ui/templates/main.html:
```diff
@@ -56,6 +56,7 @@ <h1>Φορτίο (fortio) v{{.Version}}{{if not .DoLoad}} control UI{{end}}</h1>
 Load using http: <input type="radio" name="runner" value="http" checked/> or
 grpc: <input type="radio" name="runner" value="grpc"/>
 (grpc secure transport (tls): <input type="checkbox" name="grpc-secure"/>) <br />
+CA Certificate to use when grpc secure transport (tls) is selected: <input type="text" name="ca-cert" size="64" value="/etc/ssl/certs/ca-certificates.crt" /> <br />
```
I still don't like this btw, but I can fix it myself in a follow up

ldemailly commented on fortio_main.go:
```diff
@@ -78,8 +78,10 @@ var (
 	httpsInsecureFlag = flag.Bool("https-insecure", false, "Long form of the -k flag")
+	certFlag          = flag.String("cert", "", "Path to the server certificate used for grpc TLS")
+	keyFlag           = flag.String("key", "", "Path to the server key used for grpc TLS")
+	caCertFlag        = flag.String("ca-cert", "", "Path to the CA certificate used for grpc TLS")
```
`ca-cert` -> `cacert` and add/replace at the end "for GRPC client TLS."

ldemailly commented on fortio_main.go:
```diff
@@ -78,8 +78,10 @@ var (
-	grpcSecureFlag    = flag.Bool("grpc-secure", false, "Use secure transport (tls) for GRPC")
 	httpsInsecureFlag = flag.Bool("https-insecure", false, "Long form of the -k flag")
+	certFlag          = flag.String("cert", "", "Path to the server certificate used for grpc TLS")
+	keyFlag           = flag.String("key", "", "Path to the server key used for grpc TLS")
```
"Path of the key of the TLS certificate used for GRPC server" ?

ldemailly commented on fortio_main.go:
```diff
@@ -78,8 +78,10 @@ var (
 	grpcFlag          = flag.Bool("grpc", false, "Use GRPC (health check) for load testing")
-	grpcSecureFlag    = flag.Bool("grpc-secure", false, "Use secure transport (tls) for GRPC")
 	httpsInsecureFlag = flag.Bool("https-insecure", false, "Long form of the -k flag")
+	certFlag          = flag.String("cert", "", "Path to the server certificate used for grpc TLS")
```
"Path to the certificate for TLS for the GRPC server" ?

ldemailly commented on testdata/ca.crt:
```diff
@@ -0,0 +1,30 @@
+-----BEGIN CERTIFICATE-----
```
I would make the makefile generate those rather than check in but I can do that in followup if you prefer

danehans replied:
OK. I will reintroduce [this script](https://raw.githubusercontent.com/danehans/fortio/cf67cba6ff13de698df8f7a0febb883c7871cf29/tools/cert-gen), minus the client-side creds. I will add the cert-gen script and supporting files to the project's root dir, unless you have another preference.

ldemailly replied:
that may add a dependency that isn't there though (on openssl unless there is a go way to create certs easily)

Commit:
Generates certs, adds e2e and -cacert flag

danehans commented:
Running `release-test` locally works, but it's fails in the gate:
```
++ docker run -d --ulimit nofile=20 --name fortio_server --mount type=bind,source=/go/src/istio.io/fortio/ca.crt,target=/etc/ssl/certs/ca.crt --mount type=bind,source=/go/src/istio.io/fortio/server.crt,target=/etc/ssl/certs/server.crt --mount type=bind,source=/go/src/istio.io/fortio/server.key,target=/etc/ssl/certs/server.key istio/fortio:webtest server -cert /etc/ssl/certs/server.crt -key /etc/ssl/certs/server.key -loglevel info
docker: Error response from daemon: invalid mount config for type "bind": bind source path does not exist.
```
I believe this message is the cause of the problem:
```
make[1]: Leaving directory '/go/src/istio.io/fortio'
```
The Docker -mount flag requires an absolute path. Therefore I add $PWD to mount keys/certs in `Webtest.sh`:
```
DOCKERID=$(docker run -d --ulimit nofile=$FILE_LIMIT --name $DOCKERNAME --mount type=bind,source=$PWD/ca.crt,target=$TESTCACERT --mount type=bind,source=$PWD/server.crt,target=$TESTCERT --mount type=bind,source=$PWD/server.key,target=$TESTKEY istio/fortio:webtest server -cert $TESTCERT -key $TESTKEY -loglevel $LOGLEVEL)
```

Any suggestions on how to workaround this issue?

ldemailly commented:
~while it's nice to test if mounting would work; you don't need to; just need to have make create the test certs (and put them in .gitignore); the whole fortio directory is copied into docker for stage1~

edit: nvm given the staging and that the test cert should probably not be in the release; you do need to use mount to get the test certs

danehans commented:
Partially fixes https://github.com/istio/fortio/issues/141 by allowing clients to specify a CA cert using the `-cacert` flag, but not a client cert. mTLS is required to support client certs.

danehans commented:
@ldemailly I wrote a program (`fortio/certgen/certgen.go`) to generate the ca cert and server key/cert. This should address the openssl dep issue. I am still blocked on how to get around the issue of tests passing locally, but failing in CircleCI. I believe the following message is the cause of the issue:
```
make[1]: Leaving directory '/go/src/istio.io/fortio'
```
The Docker -mount flag requires an absolute path. Therefore I add $PWD to mount keys/certs in `Webtest.sh`:
```
DOCKERID=$(docker run -d --ulimit nofile=$FILE_LIMIT --name $DOCKERNAME --mount type=bind,source=$PWD/ca.crt,target=$TESTCACERT --mount type=bind,source=$PWD/server.crt,target=$TESTCERT --mount type=bind,source=$PWD/server.key,target=$TESTKEY istio/fortio:webtest server -cert $TESTCERT -key $TESTKEY -loglevel $LOGLEVEL)
```

What is needed to have CircleCI not leave the directory `/go/src/istio.io/fortio`? Do I need to move the DOCKERID above into the Makefile?

Commit:
Removes openssl dep

ldemailly commented on certgen/certgen.go:
```diff
@@ -0,0 +1,131 @@
+)
+
+const (
+	caCrt = `
```
Sorry I wasn't clear, by generating I meant actually generating, not embedding 3 files, this isn't generating anything - it just writes 3 file content, which has no benefit over checking in files (which we don't want to do)

Anyway,  the build image does have openssl
```
$ docker run -it istio/fortio.build:v7
root@e4d71cc7d777:/go/src/istio.io# which openssl
/usr/bin/openssl
root@e4d71cc7d777:/go/src/istio.io# openssl version
OpenSSL 1.0.2g  1 Mar 2016
```
so just add a make target as I mentioned

ldemailly commented on Makefile:
```diff
@@ -25,9 +25,14 @@ GRPC_DIR:=./vendor/google.golang.org/grpc
 
+# Generate certs for unit and release tests.
+certs:
+	go run certgen/certgen.go
```
replace by openssl command *and* make it the actual file/dependency so it doesn't keep doing it if the files do exist

ldemailly commented on Makefile:
```diff
@@ -25,9 +25,14 @@ GRPC_DIR:=./vendor/google.golang.org/grpc
-test: submodule
+test: submodule certs
 	go test -timeout 60s -race $(PACKAGES)
+	rm -f ca.crt server.crt server.key
```
it shouldn't remove certs if they are there - we can add make clean-cert or some such if you wanted to force regen the certs but that's not necessary

danehans replied:
done

ldemailly commented on fgrpc/grpcrunner_test.go:
```diff
@@ -30,58 +30,271 @@ import (
 
+var (
+	// Generated from "make cert"
+	caCrt  = "../ca.crt"
```
we should probably use the mostly standard testdata/ for location of the certs (ie `fortio/fgrpc/testdata/ca.crt` etc...) but that's fine too I guess

danehans replied:
ack

ldemailly commented on ui/uihandler.go:
```diff
@@ -290,7 +291,9 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 				Destination:   url,
-				Secure:        grpcSecure,
+			}
+			if grpcSecure {
```
this should probably be just add the prefix with https:// if not already there

ldemailly commented on ui/templates/main.html:
```diff
@@ -56,6 +56,7 @@ <h1>Φορτίο (fortio) v{{.Version}}{{if not .DoLoad}} control UI{{end}}</h1>
 Load using http: <input type="radio" name="runner" value="http" checked/> or
 grpc: <input type="radio" name="runner" value="grpc"/>
 (grpc secure transport (tls): <input type="checkbox" name="grpc-secure"/>) <br />
+CA Certificate to use when grpc secure transport (tls) is selected: <input type="text" name="cacert" size="64" value="/etc/ssl/certs/ca-certificates.crt" /> <br />
```
I still think this is not useable/useful - on the UI now you have to set 2 things to make things work

danehans replied:
I am going to remove my ui changes from this PR. I will open an issue to address adding tls to the ui.

ldemailly commented on Webtest.sh:
```diff
@@ -61,11 +67,20 @@ docker exec $DOCKERNAME /usr/local/bin/fortio grpcping localhost
+# switch to secure grpc mode and run grpcping and load test
+docker stop $DOCKERID
+docker rm $DOCKERNAME
+DOCKERID=$(docker run -d --ulimit nofile=$FILE_LIMIT --name $DOCKERNAME --mount type=bind,source=$PWD/ca.crt,target=$TESTCACERT --mount type=bind,source=$PWD/server.crt,target=$TESTCERT --mount type=bind,source=$PWD/server.key,target=$TESTKEY istio/fortio:webtest server -cert $TESTCERT -key $TESTKEY -loglevel $LOGLEVEL)
```
In theory you'd mount the whole directory rather than 1 file at a time
and mount it to /etc/ssl/certs/fortio/ for instance to not shadow the existing /etc/ssl/certs which are needed for regular TLS tests
you could also then use the same image and just start another server on a different set of ports with the -cert / -key args, it'd be more efficient than a brand new docker run


**but** circleci doesn't support mounts because it's all remote docker and not local
https://circleci.com/docs/2.0/building-docker-images/#mounting-folders

you can try docker cp or we make a new layer or copy extra files/directory to  /etc/ssl/certs/fortio/ 
(that should be declared as a volume like /var/lib/istio/fortio is https://github.com/istio/fortio/blob/v0.8.1/Dockerfile#L41)

danehans replied:
thx for the ptr to the ci limitation. I have docker cp working for the cert/key files.

Commit:
Adds logs to certgen and fixes remote docker bind mount

danehans commented:
@ldemailly My next commit is ready to go and it should pass gate tests this time. I'm going to squash my commits b/c it was a huge PITA the last time I rebased.

ldemailly commented:
just update from master (pull) and it should be fine ?
if you squash or otherwise force push it's not good (for me, as it loose the reviews/state)

danehans commented:
> just update from master (pull) and it should be fine ?

that causes a merge conflict. let me see what i can do. I had rebase conflicts last time for each commit in the PR. It was painful to say the least.

Commit:
reverts to certgen script and ui

Commit:
Merge branch 'master' into grpc_certs

ldemailly commented on Makefile:
```diff
@@ -25,8 +25,12 @@ GRPC_DIR:=./vendor/google.golang.org/grpc
+
 # Local test
-test: submodule
+test: submodule certs
```
to fix the coverage build and in general : 

let's rename submodule->dependencies in every dependency clause

then make

dependencies: submodule certs

(both submodule and dependencies should be .PHONY

CERTS:=ca.crt server.crt server.key

certs: $(CERTS)

and confirm that the second time it doesn't redo the generation

danehans replied:
ack

ldemailly commented on README.md:
```diff
@@ -216,6 +228,49 @@ RTT histogram usec : count 3 avg 305.334 +/- 27.22 min 279.517 max 342.97 sum 91
+UI starting - visit:
+http://localhost:8080/fortio/
+Https redirector running on :8081
+Fortio 0.8.1 grpc ping server listening on port :8079
```
0.8.1->0.9.0 everywhere

danehans replied:
ack

ldemailly commented on cert-gen:
```diff
@@ -0,0 +1,103 @@
+# basic files/directories
+mkdir -p {certs,crl,newcerts}
+touch index.txt
+touch index.txt.attr
```
can you make a single CERT_TMP variable and directory and put every thing in that 1 directory, let's call it ./cert-tmp/ and put it in .git ignore

there are way too many files created all over right now

danehans replied:
ack

ldemailly commented on Webtest.sh:
```diff
@@ -61,14 +69,29 @@ docker exec $DOCKERNAME /usr/local/bin/fortio grpcping localhost
+# copying cert files into the certs volume
+for f in ca.crt server.crt server.key; do docker cp $PWD/$f certs:$TESTCERTSVOL/$f; done
+# start server in secure grpc mode. using non-default ports to avoid conflicts with fortio_server container.
+DOCKERSECID=$(docker run -d --ulimit nofile=$FILE_LIMIT --name $DOCKERSECNAME --volumes-from certs istio/fortio:webtest server -cert $TESTCERT -key $TESTKEY -grpc-port 8097 -http-port 8098 -redirect-port 8090 -loglevel $LOGLEVEL)
```
just need to add -cacert too to make it work for ui ;)

danehans replied:
Nope, I get error: `certificate signed by unknown authority`. It would work if we could specify a cert in the UI (which we just removed) or if cli flags were in a separate pkg, imported them into ui and assigned the cacert flag to the grpcrunner options.

ldemailly replied:
I meant add -cacert _and_ pass the value to ui.Serve but that can be a followup PR

ldemailly commented on Webtest.sh:
```diff
@@ -61,14 +69,29 @@ docker exec $DOCKERNAME /usr/local/bin/fortio grpcping localhost
 PPROF_URL="$BASE_URL/debug/pprof/heap?debug=1"
 $CURL $PPROF_URL | grep -i TotalAlloc # should find this in memory profile
+# creating dummy container to hold a volume for test certs due to remote docker bind mount limitation.
+DOCKERVOLID=$(docker create -v $TESTCERTSVOL --name certs istio/fortio.build:v6 /bin/true)
```
don't hardcode the build image if possible, also it's v7 now. it should work with the same `istio/fortio:webtest` image no (start another server with no port mapping) ?
I'm not super clear how that volumes-from works exactly ? why do we need 2 images, you can't docker cp into the second one directly - is it just a startup issue (files not being there yet?) ?

danehans replied:
I added verbosity to the comments in my upcoming commit. Let me know if you still need clarity after reviewing the inline comments.

ldemailly commented on ui/uihandler.go:
```diff
@@ -920,3 +922,14 @@ func setHostAndPort(inputPort string, addr *net.TCPAddr) {
+
+// httpToHTTPS replaces "http://" in url with "https://".
+func httpToHTTPS(url string) (parsedURL string) {
+	if strings.HasPrefix(url, "http://") {
```
no, as we discussed, this should be

addHTTPS() {
  if doesn't start with https://; add https://
}

as before I could do `istio.fortio.io:443` in the destination box and check grpc-secure and it worked and should continue to work

danehans replied:
OK. I misunderstood your original request. I have this func refactored in my upcoming commit.

Commit:
Updates makefile/Webtest/Certgen and ui https func

Commit:
Add make deps to docker compose

Commit:
Merge branch 'master' of fortio into grpc_certs

ldemailly commented on Dockerfile.test:
```diff
@@ -2,3 +2,4 @@
 FROM istio/fortio.build:v7
 WORKDIR /go/src/istio.io
 COPY . fortio
+RUN make dependencies
```
needs to be `make -C fortio dependencies`

ldemailly commented on Makefile:
```diff
@@ -14,25 +14,38 @@ TAG:=$(USER)$(shell date +%y%m%d_%H%M%S)
 	go install $(PACKAGES)
 
+# Generate certs for unit and release tests.
+certs:
```
can you make this a dependency of the files or at least 1 of the file, eg $(CERT_TEMP)/server.cert

so:
```Makefile
certs: $(CERT_TEMP)/server.cert

$(CERT_TEMP)/server.cert: cert-gen
	./cert-gen
```





so make dependencies the 2nd time doesn't do anything ? (unless cert-gen changes)

danehans replied:
ack

Commit:
Fixes make command in Dockerfile.test for CI

ldemailly commented on cert-gen:
```diff
@@ -0,0 +1,114 @@
+# Skip cert creation if the certs already exist
+if [ -d $CERT_TEMP ]
+then
+    echo "Certificate directory $CERT_TEMP exists. Skipping certificate creation."
```
ic, that is a good secondary layer
we can do the proper Makefile dependency later

danehans replied:
ack

ldemailly commented on ui/uihandler.go:
```diff
@@ -920,3 +922,18 @@ func setHostAndPort(inputPort string, addr *net.TCPAddr) {
+		log.Infof("replacing http scheme with https for url: %s", url)
+		return strings.Replace(url, "http://", "https://", 1)
+	}
+	if !strings.HasPrefix(url, "http://") && !strings.HasPrefix(url, "https://") {
```
you already tested for http

danehans replied:
ack

ldemailly commented on ui/uihandler.go:
```diff
@@ -920,3 +922,18 @@ func setHostAndPort(inputPort string, addr *net.TCPAddr) {
+func addHTTPS(url string) string {
+	if strings.HasPrefix(url, "http://") {
+		log.Infof("replacing http scheme with https for url: %s", url)
+		return strings.Replace(url, "http://", "https://", 1)
```
it's not the cheapest/cleanest way to do this (replace is a search and replace, we already know it starts with http://) but ok for now

danehans replied:
OK. I will trim if the url has a prefix of `http://` and return `https://` + the trimmed url.

Commit:
Merge branch 'master' of fortio into grpc_certs

Commit:
Added Make dep and cleaned-up addHTTPS func

Commit:
Adds add'l Dial func test cases

Commit:
Merge branch 'master' fortio into grpc_certs

danehans commented:
@ldemailly will this PR be ready to merge after I fix the conflicts?

ldemailly commented on Makefile:
```diff
@@ -14,25 +14,41 @@ TAG:=$(USER)$(shell date +%y%m%d_%H%M%S)
+
 # go test ./... and others run in vendor/ and cause problems (!)
-PACKAGES:=$(shell find . -type d -print | egrep -v "/(\.|vendor|static|templates|release|docs|json)")
+PACKAGES:=$(shell find . -type d -print | egrep -v "/(\.|vendor|static|templates|release|docs|json|cert-tmp)")
```
$(CERT_TEMP) ?

danehans replied:
removed

ldemailly commented on Makefile:
```diff
@@ -14,25 +14,41 @@ TAG:=$(USER)$(shell date +%y%m%d_%H%M%S)
 
 DOCKER_TAG = $(DOCKER_PREFIX)$(IMAGE):$(TAG)
 
+CERT_TEMP := ./cert-tmp/
```
CERT_TEMP_DIR := ...

danehans replied:
done

ldemailly commented on README.md:
```diff
@@ -216,6 +228,49 @@ RTT histogram usec : count 3 avg 305.334 +/- 27.22 min 279.517 max 342.97 sum 91
+UI starting - visit:
+http://localhost:8080/fortio/
+Https redirector running on :8081
+Fortio 0.9.0 grpc ping server listening on port :8079
```
when merging make sure there is 0.9.1 everywhere

danehans replied:
Looks like docs are now using 0.10.0. I have changed to 0.10.0.

ldemailly replied:
it's actually 0.10.1 now :-)

ldemailly commented on README.md:
```diff
@@ -216,6 +228,49 @@ RTT histogram usec : count 3 avg 305.334 +/- 27.22 min 279.517 max 342.97 sum 91
+* A `grpcping` using TLS. First, start Fortio server with the `-cert` and `-key` flags.
+`/path/to/fortio/server.crt` and `/path/to/fortio/server.key` are paths to the TLS certificate and key that
+you are responsible for providing. Creating a TLS certificate is outside the scope of this document. This example uses
+`localhost` in the `Subject Alternative Name` (SAN) of the certificate. The SAN should match the domain name that
```
remove that part (localhost san) as it doesn't matter / is also out of scope ? - or mentioned the makefile can generate test only certs

danehans replied:
done

ldemailly commented on README.md:
````diff
@@ -336,6 +391,14 @@ And the JSON saved is
+Next, run the `load` command with the `-cacert` flag:
+```
+$ fortio load -cacert /etc/ssl/certs/ca.crt -grpc localhost:8079
+
````
missing close 3 ` - please make sure you preview/look over the changes

danehans replied:
done

ldemailly commented on Webtest.sh:
```diff
@@ -6,7 +6,12 @@ FORTIO_UI_PREFIX=/newprefix/ # test the non default prefix (not /fortio/)
+CERT=/etc/ssl/certs/ca-certificates.crt
+TEST_CERT_VOL=/etc/ssl/certs/fortio
+BUILD_IMAGE_TAG=v7
+BUILD_IMAGE=istio/fortio.build:$BUILD_IMAGE_TAG
```
change so it works with https://github.com/istio/fortio/blob/master/Makefile#L112

ie: add the file name to the list and use istio/fortio.build:v7 directly so sed finds it when we go to v8

ldemailly commented on fortio_main.go:
```diff
@@ -78,8 +78,10 @@ var (
 	goMaxProcsFlag    = flag.Int("gomaxprocs", 0, "Setting for runtime.GOMAXPROCS, <1 doesn't change the default")
 	profileFlag       = flag.String("profile", "", "write .cpu and .mem profiles to file")
 	grpcFlag          = flag.Bool("grpc", false, "Use GRPC (health check by default, add -ping for ping) for load testing")
-	grpcSecureFlag    = flag.Bool("grpc-secure", false, "Use secure transport (tls) for GRPC")
```
that change isn't reflected in the README.md

see release/updateFlags.sh for how to update

we also need to document somewhere that https:// replaces -grpc-secure

danehans replied:
I have updated the readme to remove the -secure-grpc flag and ran the updateFlags.sh script.

danehans replied:
done

ldemailly reviewed (commented):
almost, see below

danehans commented:
@ldemailly thanks for the feedback. I need a couple days to wrap-up something I'm working on and I'll push a new commit that addresses your concerns.

ldemailly commented:
thanks for your patience and flexibility !

Commit:
Merge branch 'master' fortio into grpc_certs

Commit:
Addresses @ldemailly comments

danehans reviewed (commented):
I'm getting ready to push my latest commit that addresses your feedback.

danehans reviewed (commented):
I'm getting ready to push my latest commit that addresses your feedback.

Commit:
Merge branch 'master' of fortio into grpc_certs

Commit:
Adds TestExitedPingServer test

Commit:
Refactor PingServer test for failed crt/key

danehans commented:
@ldemailly do you mind reviewing when you have a moment?

ldemailly commented:
looking now

ldemailly commented:
getting some github (unicorn) errors when I try to open this PR somehow)

qq: why did you remove  TestExitedPingServer ?

ldemailly committed:
Merge branch 'master' into grpc_certs

ldemailly committed:
use 0.10.1 like rest of readme

ldemailly commented on Webtest.sh:
```diff
@@ -61,14 +64,30 @@ docker exec $DOCKERNAME /usr/local/bin/fortio grpcping localhost
 docker stop $DOCKERID
 docker rm $DOCKERNAME
+docker stop $DOCKERSECID
+docker rm $DOCKERSECNAME
```
this should probably be "trap"ed too but I guess we can clean that up later

danehans replied:
I thought the trap is used for the final Fortio instance, the report mode server. I stop/rm the Docker assets for the secure grpc instance of Fortio server before the script starts the report mode Fortio server.

ldemailly replied:
trap means cleanup like a finally, in case of error in any of tbe tests

ldemailly commented on fgrpc/grpcrunner.go:
```diff
@@ -43,17 +43,30 @@ const (
+	conn, err = grpc.Dial(serverAddr, opts...)
 	if err != nil {
-		log.Errf("failed to conect to %s with tls %v: %v", serverAddr, tls, err)
+		log.Errf("failed to connect to %s: %v", serverAddr, err)
```
NeedFix: when an error occurs it is very important to log all the input, please add cacert and override to the statement

danehans replied:
ack

ldemailly commented on fgrpc/grpcrunner.go:
```diff
@@ -43,17 +43,30 @@ const (
+// Dial dials grpc using insecure or tls transport security when serverAddr
+// has prefixHTTPS or cert is provided. If override is set to a non empty string,
+// it will override the virtual host name of authority in requests.
+func Dial(serverAddr string, cacert, override string) (conn *grpc.ClientConn, err error) {
```
NeedFix: either put the type for all 3 or only once but here it's odd to have string 2 out of 3 args

danehans replied:
ack

ldemailly commented on fgrpc/grpcrunner_test.go:
```diff
@@ -106,31 +240,127 @@ func TestGRPCRunnerMaxStreams(t *testing.T) {
+				},
+				Destination:  sDest,
+				CACert:       caCrt,
+				CertOverride: "invalidName",
```
can we have a positive test of CertOverride (where it is set yet works ?)

ldemailly commented on fgrpc/grpcrunner_test.go:
```diff
@@ -30,34 +30,168 @@ import (
+		{
+			name: "insecure runner with payload",
+			runnerOpts: GRPCRunnerOptions{
+				RunnerOptions: periodic.RunnerOptions{
```
NeedFix: there is a lot of copy paste of this section, why not have the runner options in a variable ?

danehans replied:
I guess I find table tests easier to read when testing multiple test cases. WOuld you rather I follow the existing pattern for the test? Something like this:

```
func TestGRPCRunner(t *testing.T) {
	log.SetLogLevel(log.Info)
	iPort := PingServer("0", "", "", "bar", 0)
	iDest := fmt.Sprintf("localhost:%d", iPort)
	sPort := PingServer("0", svrCrt, svrKey, "bar", 0)
	sDest := fmt.Sprintf("localhost:%d", sPort)
	ro := periodic.RunnerOptions{
		QPS:        100,
		Resolution: 0.00001,
	}

	opts := GRPCRunnerOptions{
		RunnerOptions: ro,
		Destination: iDest,
		Profiler:    "test.profile",
	}
	res, err := RunGRPCTest(&opts)
	if err != nil {
		t.Error(err)
		return
	}
	totalReq := res.DurationHistogram.Count
	ok := res.RetCodes[grpc_health_v1.HealthCheckResponse_SERVING]
	if totalReq != ok {
		t.Errorf("Mismatch between requests %d and ok %v", totalReq, res.RetCodes)
	}
	// Test a grpc runner with TLS enabled.
	opts.Destination = sDest
	opts.CACert = caCrt
	res, err = RunGRPCTest(&opts)
	if err != nil {
		t.Error(err)
		return
	}
	totalReq = res.DurationHistogram.Count
	ok = res.RetCodes[grpc_health_v1.HealthCheckResponse_SERVING]
	if totalReq != ok {
		t.Errorf("Mismatch between requests %d and ok %v", totalReq, res.RetCodes)
	}
}
```

Keep in mind the above example is only exercising 2 test cases.

ldemailly commented on fgrpc/grpcrunner_test.go:
```diff
@@ -106,31 +240,127 @@ func TestGRPCRunnerMaxStreams(t *testing.T) {
+		{
+			name: "insecure runner",
+			runnerOpts: GRPCRunnerOptions{
+				RunnerOptions: periodic.RunnerOptions{
```
NeedFix: same

ldemailly commented on fgrpc/pingsrv.go:
```diff
@@ -66,6 +67,15 @@ func PingServer(port string, healthServiceName string, maxConcurrentStreams uint
+			log.Fatalf("Invalid TLS credentials: %v\n", err)
+		}
+		log.Infof("Using server certificate %v to construct TLS credentials", cert)
+		log.Infof("Using server key %v to construct TLS credentials", key)
```
this could probably be a single log info instead of 2 but it's ok as is too

ldemailly reviewed (commented):
Really almost there now - Thanks! 

A few comment and you only need to update where I prefixed by "NeedFix" (4 places)

danehans reviewed (commented):
Please see my feedback and example refactor of `TestGRPCRunner`. I would like to make sure we're on the same page before I refactor `TestGRPCRunner` and `TestGRPCRunnerWithError`.

danehans reviewed (commented):
Please see my feedback and example refactor of `TestGRPCRunner`. I would like to make sure we're on the same page before I refactor `TestGRPCRunner` and `TestGRPCRunnerWithError`.

Commit:
Addresses ldemailly feedback

Commit:
Merge branch 'master' of fortio into grpc_certs

Commit:
Merge branch 'grpc_certs' fortio into grpc_certs

Commit:
Removed docker stop/rm from trap

ldemailly commented on Webtest.sh:
```diff
@@ -14,6 +14,9 @@ DOCKERID=$(docker run -d --ulimit nofile=$FILE_LIMIT --name $DOCKERNAME istio/fo
 function cleanup {
   docker stop $DOCKERID
   docker rm $DOCKERNAME
+  docker stop $DOCKERSECID
```
in theory should have an if that it is set/started (as it's not started right away unlike DOCKERID) but it's ok, you didn't have to change it (for that reason as well - I'll try - one test is interrupt or have a test error out and see if it still finishes correctly/cleansup without too many additional errors)

ldemailly commented on fgrpc/grpcrunner.go:
```diff
@@ -43,17 +43,30 @@ const (
+			log.Errf("Invalid TLS credentials: %v\n", err)
+			return nil, err
+		}
+		log.Infof("Using CA certificate %v to construct TLS credentials", cacert)
```
optional: here too (override logging)

ldemailly commented on fgrpc/grpcrunner_test.go:
```diff
@@ -30,34 +30,133 @@ import (
-	if totalReq != ok {
-		t.Errorf("Mismatch between requests %d and ok %v", totalReq, res.RetCodes)
+	for _, test := range tests {
+		switch {
```
I didn't mean to have to put a big switch here :( why can't those be above directly ?

ldemailly commented:
closing for #249 (this PR makes github choke so redone there)
to see history/comments check with incognito window (and/or when github fixes that bug)
